File: Add Code Here/JAVA/LeetCode-Solutions-Java/102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> list = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        if(root == null) {
            return list;
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        q.add(null);
        while(!q.isEmpty()) {
            TreeNode curNode = q.remove();
            if(curNode == null) {
                list.add(temp);
                temp = new ArrayList<>();
                if(q.isEmpty()) {
                    break;
                }
                else {
                    q.add(null);
                }
            }
            else {
                temp.add(curNode.val);
                if(curNode.left != null) {
                    q.add(curNode.left);
                }
                if(curNode.right != null) {
                    q.add(curNode.right);
                }
            }
        }
        return list;
    }
}

File: Add Code Here/C++/tree/c++ problems asked in interview/102. Binary Tree Level Order Traversal.cpp
Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

Example 1:

Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]
Example 2:

Input: root = [1]
Output: [[1]]
Example 3:

Input: root = []
Output: []

Solution :
Time Complexity - O(n) where n is the number of nodes in the binary tree
Space Complexity - O(n) because we are using a vector to store the traversal level wise

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>>ans;
        if(root==NULL) return ans; //incase the root has no value
        
        queue<TreeNode*>q; // queue ds 
        q.push(root);      // pushing the initial root value into the queue
        while(!q.empty())  //if queue is not empty 
        {
            int size = q.size(); // find the size of the queue , which means it gives the size of no. of elements in the queue
            vector<int>level;    // a vector to store level wise traversals 
            for(int i=0;i<size;i++)
            { 
                TreeNode* node = q.front();  //pointer to the front element of the queue 
                q.pop();                      //pop the front element 
                if(node->left!=NULL)          // if node is not null, push the values in the queue 
                    q.push(node->left);
                if(node->right!=NULL)
                    q.push(node->right);
                level.push_back(node->val);   // pushing the values of the node in the level vector 
            }
            ans.push_back(level);          // pushing the level vector traversal in the ans vector 
        }
        return ans;                           // returning the answer 
    }
};
